### ISSUE TITLE
Error downloading module: "host is not an insecure source"

### ISSUE THREAD

**Issue by dlorenc** [kubewarden/policy-server#59]
I tried to deploy the scaffold project on GCR/GKE, and ran into this error:

```
May 08 17:43:05.828 ERROR policy_server: error while fetching policy name-check from registry://us.gcr.io/dlorenc-vmtest2/wasm: could not download Wasm module: error decoding response body: missing field `detail` at line 1 column 137; host is not an insecure source
```

The registry itself is public and does not require auth to pull. It looks like from reading the code I might need to set something in a sources.yml, but I'm not sure what or where to do that.

**Comment by ereslibre** [comment]
Thank you for your report @dlorenc. I'm going to check the root cause of this issue as I've seen this myself with `kwctl`.

**Comment by dlorenc** [comment]
Thanks! It appears to work for me with ghcr.io, but not gcr.io

**Comment by ereslibre** [comment]
The error is wrong. The sources.yaml file would be needed if you had a custom CA or if you wanted to skip TLS verification of the host.

I cannot look too much into it right now, but I wonder if you can pull the Wasm module by using `wasm-to-oci pull` from your CLI. I tried `wasm-to-oci pull us.gcr.io/dlorenc-vmtest2/wasm` but got an `Error: object required` from `wasm-to-oci`.

Also, given it's a registry I wonder if we should default to `:latest` tag if no tag is provided, as in your case. Many of these behaviors that we take for granted on Docker and such has to be revisited here. I tried `wasm-to-oci pull us.gcr.io/dlorenc-vmtest2/wasm:latest` and I got a 404. Is it desired to be able to pull from a registry without a tag at all?

**Comment by dlorenc** [comment]
Thanks! This does look like a strange bug with wasm-to-oci. I'll post here for posterit:

It works with no tag:

```shell
$ wasm-to-oci push policy.wasm us.gcr.io/dlorenc-vmtest2/wasm
INFO[0000] Pushed: us.gcr.io/dlorenc-vmtest2/wasm
INFO[0000] Size: 185615
INFO[0000] Digest: sha256:d92cfb03e5f91bfacdf00c96bf33998de20bf697269d9c9930e4f8fd5de54d62
```

But it doesn't work with a tag:

```shell
wasm-to-oci push policy.wasm us.gcr.io/dlorenc-vmtest2/wasm:foo
Error: failed commit on ref "manifest-sha256:d92cfb03e5f91bfacdf00c96bf33998de20bf697269d9c9930e4f8fd5de54d62": unexpected status: 401 Unauthorized
```

After the push with no tag, I can't find it:

```shell
$ wasm-to-oci pull us.gcr.io/dlorenc-vmtest2/wasm
Error: object required
```

Or with crane:

```shell
crane manifest us.gcr.io/dlorenc-vmtest2/wasm
2021/05/08 14:37:54 fetching manifest us.gcr.io/dlorenc-vmtest2/wasm: GET https://us.gcr.io/v2/dlorenc-vmtest2/wasm/manifests/latest: MANIFEST_UNKNOWN: Failed to fetch "latest" from request "/v2/dlorenc-vmtest2/wasm/manifests/latest".
```

Thanks for the help! I'll see if I can figure out what's wrong with wasm-to-oci!

**Comment by ereslibre** [comment]
Thanks a lot @dlorenc, feel free to let us know what you find, we are also interested in improving the UX and this use cases. I'll take care of fixing the error we are reporting that is misleading. In that regard your report is perfectly fine! That's why I'm reopening it until I take a look.

**Comment by dlorenc** [comment]
I got it working with cosign and go-containerregistry here: https://github.com/sigstore/cosign/pull/315

I couldn't get wasm-to-oci working with gcr, I'll keep looking into that.

**Comment by ereslibre** [comment]
That's super cool @dlorenc!

Just a comment: `wasm-to-oci` was just a way to rule out issues with connectivity or otherwise pull access of the Wasm module. The `policy-server` and `kwctl` projects both depend on the [`policy-fetcher`](https://github.com/kubewarden/policy-fetcher) crate, that depends on the [`oci-distribution`](https://github.com/deislabs/krustlet/tree/main/crates/oci-distribution) crate from the krustlet project. So, errors shown on the policy-server and kwctl are coming from that crate.

Also, as mentioned earlier, we have an issue with the error reporting by the policy-fetcher crate that is misleading (the error you got).

**Comment by ereslibre** [comment]
@dlorenc I have opened https://github.com/deislabs/krustlet/pull/593 to fix part of the issue.

The error coming from the `us.gcr.io` does not contain a `detail` field. However, the `oci-distribution` crate was trying to unmarshal the whole struct. This is the explanation for the `error decoding response body: missing field `detail` at line 1 column 137` bit in your error. By making it defaulting automatically, we don't get that error anymore. It reads now:

```
ereslibre@desktop ~/p/k/kwctl (main)> cargo run -- pull registry://us.gcr.io/dlorenc-vmtest2/wasm
...
Error: could not download Wasm module: OCI API error: Failed to fetch "latest" from request "/v2/dlorenc-vmtest2/wasm/manifests/latest". on https://us.gcr.io/v2/dlorenc-vmtest2/wasm/manifests/latest; host is not an insecure source
```

**Comment by flavio** [comment]
We are now building from a fork of `oci-distribution` that includes @ereslibre's fix. This issue should be solved with policy-server v0.1.5.
I don't have access to a gcr to verify the fix.

@dlorenc: can you still reproduce the issue?

**Comment by ereslibre** [comment]
I cannot reproduce this issue with latest kwctl. Thank you for your report @dlorenc.
